### PR TITLE
Remove the 'with' statement from the blackbird grammar, and replace it with a metadata block

### PR DESCRIPTION
**Pull request by josh146**
The new blackbird script for teleportation under this change:

```python
# required metadata at the top of the file
name StateTeleportation
# required blackbird/cloud API version it targets
version 1.0
# optional 'target' specifies the device the blackbird
# program was compiled for, as well as any other device
# metadata/options
target gaussian (shots=1000)

# state to be teleported:
complex alpha = 1+0.5j
Coherent(alpha) | 0

# maximally entangled states
Squeezed(-4) | 1
Squeezed(4) | 2
BSgate(pi/4, 0) | (1, 2)

# Alice performs the joint measurement
# in the maximally entangled basis
int b = 5
BSgate(pi/4, b) | (0, 1)
MeasureX | 0
MeasureP | 1

# Bob conditionally displaces his mode
# based on Alice's measurement result
Xgate(sqrt(2)*q0) | 2
Zgate(sqrt(2)*q1) | 2

MeasureHeterodyne() | 2
```

### CODE REVIEW

**Comment by josh146** [comment]
As before, all files matching the pattern `blackbird*.*` are autogenerated and can be ignored

File: apps/strawberry_fields_listener.py [hunk 1]
```diff
@@ -35,16 +35,15 @@ def __init__(self):
...
     def run(self):
-        if 'num_subsystems' not in self.device['kwargs']:
-            raise ValueError("Must specify the number of subsystems")
+        if self.target is None:
+            raise ValueError("Blackbird program has no target backend")
 
         self.eng, self.q = sf.Engine(
-            self.device['kwargs']['num_subsystems'],
-            hbar=self.device['kwargs'].get('hbar', 2)
+            len(self.active_modes),
```
Comment (co9olguy) [inline apps/strawberry_fields_listener.py]: Is this stuff an unrelated change that snuck into the same PR?

File: doc/apps/sf.rst [hunk 2]
```diff
@@ -62,13 +65,13 @@ The program should produce the following output:
...
-    [-0.38611559831595721, 0.87013137205881896, -0.8637591428902387, 0.87987666305358891]
-    [-0.90073440739802579, 0.6541730982429812, -0.17966121619754322, 0.7105191725142429]
-    [2.2181871945635283, 0.90784717914877755, -0.14608293659572136, 1.3234082450308977]
-    [0.26187400468879646, -0.24350554600924179, 0.15135585268536278, -1.3698092903244679]
-    [0.62928125081235919, -2.1680158143971759, -0.99450040689765284, 1.8511289378967497]
-    [1.9602717053077034, -1.8998580204278686, 0.23193528358577153, 1.5632428346967497]
-    [0.34002902283984326, 0.99624604317845789, -0.32804961033646657, 0.89742913874610075]
-    [-0.58834092978787944, -1.0543538025898438, 0.18153026050309107, 0.81751881123822168]
-    [1.3795212337550613, -1.0187713250385, 0.50738610139349849, 0.25778266780662962]
+    [-0.17006682173581544, -0.09825970340668835, 0.44794947054926926, -0.052141596937362866]
```
Comment (co9olguy) [inline doc/apps/sf.rst]: Different results? It doesn't look like the program changed... :thinking:

File: doc/syntax.rst [hunk 3]
```diff
@@ -221,29 +245,28 @@ columns separated by commas.
 Quantum program
 ---------------
 
-The ``with`` statement indicates the device to run the program on,
-as well as providing device-specific options.
+The ``|`` operator is used to apply intrinsic quantum operations to quantum registers.
 
-Inside the indented ``with`` block, all operations are queued
-to be executed on the device, in the order they appear.
+All operations are queued to be executed on the device, in the order they appear.
```
Comment (co9olguy) [inline doc/syntax.rst]: Is this something that needs to be said in the blackbird docs? i.e., is it more appropriate elsewhere?

File: examples/example.xbb [hunk 4]
```diff
@@ -16,14 +16,31 @@
...
+# -------------------------------------------
+# At the top of the file, we define:
+#
+# * name: the name of the quantum program,
+# * version: the Blackbird version the script targets
+# * target: (optional) device/simulator target
+#   that the quantum program has been compiled for.
+#
+# In addition, keyword options can be provided for the
+# target using the syntax (option1=0.34, option2=True)
```
Comment (co9olguy) [inline examples/example.xbb]:
```suggestion
# target using the syntax `(option1=0.34, option2=True)`.
```
Not sure if this is better, but I think we want to highlight which part is code

File: examples/example.xbb [hunk 5]
```diff
@@ -16,14 +16,31 @@
...
+name example
+version 1.0
+
+target Chip0 (shots=10)
+
 # -------------------------------------------
 # Variable declarations
 # -------------------------------------------
-# Variable are defined at the top of the file.
+# Variable can be defined anywhere in the file.
```
Comment (co9olguy) [inline examples/example.xbb]:
```suggestion
# Variables can be defined anywhere in the file.
```

File: examples/example.xbb [hunk 6]
```diff
@@ -71,64 +89,64 @@ complex array A =
...
+Dgate(0.43, 0.12) | 3
+Dgate(1+2j) | 1
+Sgate(0.43, -0.123) | 2
+Sgate(0.432) | 3
+Rgate(pi) | 2
+Fouriergate | 0
 
-    Catstate(0.4+0.1j) | 0
-    Catstate(0.4+0.1j, 1.0) | 0
+# Multiple modes are specified by comma separated integers
```
Comment (co9olguy) [inline examples/example.xbb]:
```suggestion
# Multiple modes are specified by comma-separated integers
```

File: examples/state_teleportation.xbb [hunk 7]
```diff
@@ -1,37 +1,31 @@
...
-
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License.
-
+# required metadata at the top of the file
+name StateTeleportation
+# required blackbird/cloud API version it targets
```
Comment (co9olguy) [inline examples/state_teleportation.xbb]:
```suggestion
# required blackbird API version it targets
```

File: apps/strawberry_fields_listener.py (hunk 1 above)
Comment (josh146) [reply apps/strawberry_fields_listener.py]: Slightly related. Since we no longer store `eng.run()` compatible arguments in the blackbird file, I had to modify this to 'extract' the num_subsystems from the blackbird script, to keep it working.

File: doc/apps/sf.rst (hunk 2 above)
Comment (josh146) [reply doc/apps/sf.rst]: These are measurement results, so it's different every time the program runs. I'm not sure why I updated it in the docs though :thinking: Maybe I was being overly cautious.

File: doc/syntax.rst (hunk 3 above)
Comment (josh146) [reply doc/syntax.rst]: No, I can remove this.
```suggestion
```

File: examples/example.xbb (hunk 4 above)
Comment (josh146) [reply examples/example.xbb]: Sure. Using `(option1=0.34, option2=True)` is a stopgap solution, since kwargs were already supported by the grammar.

If you prefer something else, I can do that in another PR. Another option I considered:

```python
target chip0
    - shots=100
```

But this is a minor change, so I decided it was out of scope of this PR for now.

File: doc/apps/sf.rst (hunk 2 above)
Comment (co9olguy) [reply doc/apps/sf.rst]: lol if they change every time, then why put the output in the docs? People will think they are getting the wrong answer if they ever try to run it

File: apps/strawberry_fields_listener.py (hunk 1 above)
Comment (co9olguy) [reply apps/strawberry_fields_listener.py]: Looks better the new way

File: doc/syntax.rst (hunk 3 above)
Comment (co9olguy) [reply doc/syntax.rst]: :+1:

File: examples/example.xbb (hunk 4 above)
Comment (co9olguy) [reply examples/example.xbb]: No, I just meant that in the comment, we should put quotes to emphasize which parts are code vs comments

File: doc/apps/sf.rst (hunk 2 above)
Comment (josh146) [reply doc/apps/sf.rst]: Good point lol

Comment (josh146) [reply doc/apps/sf.rst]: Clarified with a note in the latest commit.
